File: lib/extraction/arithmetic.ts
import type { ControlResult, ExtractedInvoice } from "@/lib/types";
import { addCents, centsDiff, centsToDecimalString, parseDecimalToCents } from "@/lib/money";

export type ArithmeticPolicy = {
  taxRoundingToleranceUsd: number;
};

/**
 * Recomputes every arithmetic relationship in the invoice using integer-cent
 * math — never trusting the printed figures. Produces the same 3 control
 * rows the Phase 1 fixtures hand-authored (arithmetic_line_totals,
 * arithmetic_subtotal, arithmetic_tax_total), now computed for real.
 *
 * A line whose quantity/unitPrice/lineTotal isn't independently verified
 * against the document (see align-evidence.ts) cannot be arithmetic-checked
 * at all — that's reported as its own failed, blocking control rather than
 * silently skipped, which is what actually enforces "no uncertain required
 * monetary field passes automatically."
 */
export function computeArithmeticControls(
  extracted: ExtractedInvoice,
  policy: ArithmeticPolicy
): ControlResult[] {
  const controls: ControlResult[] = [];
  const toleranceCents = Math.round(policy.taxRoundingToleranceUsd * 100);

  // --- Line-total recalculation ---
  const lineCents: (number | null)[] = [];
  const lineIssues: string[] = [];

  for (const li of extracted.lineItems) {
    const unresolved =
      li.quantity.status !== "verified" ||
      li.unitPrice.status !== "verified" ||
      li.lineTotal.status !== "verified";

    if (unresolved) {
      lineCents.push(null);
      lineIssues.push(
        `Line ${li.lineNumber} (${li.description.value ?? "unknown"}): quantity, unit price, or line total could not be independently verified against the document — cannot recompute.`
      );
      continue;
    }

    const qty = Number(li.quantity.value);
    const unitCents = parseDecimalToCents(li.unitPrice.value);
    const printedLineCents = parseDecimalToCents(li.lineTotal.value);

    if (!Number.isFinite(qty) || unitCents === null || printedLineCents === null) {
      lineCents.push(null);
      lineIssues.push(`Line ${li.lineNumber}: quantity, unit price, or line total is not a valid number.`);
      continue;
    }

    const computedLineCents = Math.round(qty * unitCents);
    lineCents.push(printedLineCents);

    if (centsDiff(computedLineCents, printedLineCents) > 1) {
      lineIssues.push(
        `Line ${li.lineNumber} (${li.description.value}): ${qty} × $${li.unitPrice.value} = $${centsToDecimalString(computedLineCents)}, but the printed line total is $${li.lineTotal.value}.`
      );
    }
  }

  controls.push({
    controlId: "arithmetic_line_totals",
    label: "Line-total recalculation",
    status: lineIssues.length === 0 ? "passed" : "failed",
    severity: lineIssues.length === 0 ? "low" : "high",
    reason:
      lineIssues.length === 0
        ? `All ${extracted.lineItems.length} line total(s) recompute exactly from quantity × unit price.`
        : lineIssues.join(" "),
    evidenceReferences: ["lineItems"],
    blocking: true,
  });

  // --- Subtotal recalculation ---
  const allLinesResolved = lineCents.every((c) => c !== null);
  const summedLineCents = allLinesResolved ? addCents(...(lineCents as number[])) : null;
  const printedSubtotalCents =
    extracted.subtotal.status === "verified" ? parseDecimalToCents(extracted.subtotal.value) : null;

  let subtotalStatus: ControlResult["status"] = "failed";
  let subtotalReason: string;
  if (!allLinesResolved) {
    subtotalReason = "Cannot recompute the subtotal — one or more line totals were not independently verified.";
  } else if (printedSubtotalCents === null) {
    subtotalReason = "Subtotal field is not a verified, valid monetary value — cannot check against line totals.";
  } else if (centsDiff(summedLineCents as number, printedSubtotalCents) <= 1) {
    subtotalStatus = "passed";
    subtotalReason = `Sum of line totals equals the printed subtotal of $${extracted.subtotal.value}.`;
  } else {
    subtotalReason = `Sum of line totals is $${centsToDecimalString(summedLineCents as number)}, but the printed subtotal is $${extracted.subtotal.value}.`;
  }

  controls.push({
    controlId: "arithmetic_subtotal",
    label: "Subtotal recalculation",
    status: subtotalStatus,
    severity: subtotalStatus === "passed" ? "low" : "high",
    reason: subtotalReason,
    evidenceReferences: ["subtotal"],
    blocking: true,
  });

  // --- Tax and grand-total recalculation ---
  const taxCents = extracted.tax.status === "verified" ? parseDecimalToCents(extracted.tax.value) : null;
  const totalCents = extracted.total.status === "verified" ? parseDecimalToCents(extracted.total.value) : null;

  let taxStatus: ControlResult["status"] = "failed";
  let taxReason: string;
  if (printedSubtotalCents === null || taxCents === null || totalCents === null) {
    taxReason = "Subtotal, tax, or total is not a verified, valid monetary value — cannot recompute.";
  } else {
    const computedTotalCents = addCents(printedSubtotalCents, taxCents);
    const diff = centsDiff(computedTotalCents, totalCents);
    if (diff <= toleranceCents) {
      taxStatus = "passed";
      taxReason =
        diff === 0
          ? `Subtotal + tax equals the printed total of $${extracted.total.value} exactly.`
          : `Subtotal + tax = $${centsToDecimalString(computedTotalCents)}; printed total is $${extracted.total.value} — $${centsToDecimalString(diff)} difference, within the $${policy.taxRoundingToleranceUsd.toFixed(2)} rounding tolerance.`;
    } else {
      taxReason = `Subtotal ($${extracted.subtotal.value}) + tax ($${extracted.tax.value}) = $${centsToDecimalString(computedTotalCents)}, but the printed total is $${extracted.total.value} — exceeds the $${policy.taxRoundingToleranceUsd.toFixed(2)} rounding tolerance.`;
    }
  }

  controls.push({
    controlId: "arithmetic_tax_total",
    label: "Tax and grand-total recalculation",
    status: taxStatus,
    severity: taxStatus === "passed" ? "low" : "high",
    reason: taxReason,
    evidenceReferences: ["tax", "total"],
    blocking: true,
  });

  return controls;
}

File: ledgerguard/extraction/arithmetic.py
"""Recomputes every arithmetic relationship in the invoice using integer-cent
math — never trusting the printed figures."""

from ..money import add_cents, cents_diff, cents_to_decimal_string, parse_decimal_to_cents


def compute_arithmetic_controls(extracted: dict, tax_rounding_tolerance_usd: float) -> list[dict]:
    """A line whose quantity/unitPrice/lineTotal isn't independently verified
    against the document cannot be arithmetic-checked at all — that's
    reported as its own failed, blocking control rather than silently
    skipped, which is what actually enforces "no uncertain required monetary
    field passes automatically."""
    controls: list[dict] = []
    tolerance_cents = round(tax_rounding_tolerance_usd * 100)

    # --- Line-total recalculation ---
    line_cents: list[int | None] = []
    line_issues: list[str] = []

    for li in extracted["lineItems"]:
        unresolved = (
            li["quantity"].status != "verified"
            or li["unitPrice"].status != "verified"
            or li["lineTotal"].status != "verified"
        )

        if unresolved:
            line_cents.append(None)
            desc = li["description"].value or "unknown"
            line_issues.append(
                f"Line {li['lineNumber']} ({desc}): quantity, unit price, or line total could not be "
                "independently verified against the document — cannot recompute."
            )
            continue

        try:
            qty = float(li["quantity"].value)
        except (TypeError, ValueError):
            qty = None
        unit_cents = parse_decimal_to_cents(li["unitPrice"].value)
        printed_line_cents = parse_decimal_to_cents(li["lineTotal"].value)

        if qty is None or unit_cents is None or printed_line_cents is None:
            line_cents.append(None)
            line_issues.append(f"Line {li['lineNumber']}: quantity, unit price, or line total is not a valid number.")
            continue

        computed_line_cents = round(qty * unit_cents)
        line_cents.append(printed_line_cents)

        if cents_diff(computed_line_cents, printed_line_cents) > 1:
            line_issues.append(
                f"Line {li['lineNumber']} ({li['description'].value}): {qty} × "
                f"${li['unitPrice'].value} = ${cents_to_decimal_string(computed_line_cents)}, but the "
                f"printed line total is ${li['lineTotal'].value}."
            )

    controls.append(
        {
            "controlId": "arithmetic_line_totals",
            "label": "Line-total recalculation",
            "status": "passed" if not line_issues else "failed",
            "severity": "low" if not line_issues else "high",
            "reason": (
                f"All {len(extracted['lineItems'])} line total(s) recompute exactly from quantity × unit price."
                if not line_issues
                else " ".join(line_issues)
            ),
            "evidenceReferences": ["lineItems"],
            "blocking": True,
        }
    )

    # --- Subtotal recalculation ---
    all_lines_resolved = all(c is not None for c in line_cents)
    summed_line_cents = add_cents(*[c for c in line_cents if c is not None]) if all_lines_resolved else None
    printed_subtotal_cents = (
        parse_decimal_to_cents(extracted["subtotal"].value) if extracted["subtotal"].status == "verified" else None
    )

    subtotal_status = "failed"
    if not all_lines_resolved:
        subtotal_reason = "Cannot recompute the subtotal — one or more line totals were not independently verified."
    elif printed_subtotal_cents is None:
        subtotal_reason = "Subtotal field is not a verified, valid monetary value — cannot check against line totals."
    elif cents_diff(summed_line_cents, printed_subtotal_cents) <= 1:
        subtotal_status = "passed"
        subtotal_reason = f"Sum of line totals equals the printed subtotal of ${extracted['subtotal'].value}."
    else:
        subtotal_reason = (
            f"Sum of line totals is ${cents_to_decimal_string(summed_line_cents)}, but the printed "
            f"subtotal is ${extracted['subtotal'].value}."
        )

    controls.append(
        {
            "controlId": "arithmetic_subtotal",
            "label": "Subtotal recalculation",
            "status": subtotal_status,
            "severity": "low" if subtotal_status == "passed" else "high",
            "reason": subtotal_reason,
            "evidenceReferences": ["subtotal"],
            "blocking": True,
        }
    )

    # --- Tax and grand-total recalculation ---
    tax_cents = parse_decimal_to_cents(extracted["tax"].value) if extracted["tax"].status == "verified" else None
    total_cents = parse_decimal_to_cents(extracted["total"].value) if extracted["total"].status == "verified" else None

    tax_status = "failed"
    if printed_subtotal_cents is None or tax_cents is None or total_cents is None:
        tax_reason = "Subtotal, tax, or total is not a verified, valid monetary value — cannot recompute."
    else:
        computed_total_cents = add_cents(printed_subtotal_cents, tax_cents)
        diff = cents_diff(computed_total_cents, total_cents)
        if diff <= tolerance_cents:
            tax_status = "passed"
            if diff == 0:
                tax_reason = f"Subtotal + tax equals the printed total of ${extracted['total'].value} exactly."
            else:
                tax_reason = (
                    f"Subtotal + tax = ${cents_to_decimal_string(computed_total_cents)}; printed total is "
                    f"${extracted['total'].value} — ${cents_to_decimal_string(diff)} difference, within the "
                    f"${tax_rounding_tolerance_usd:.2f} rounding tolerance."
                )
        else:
            tax_reason = (
                f"Subtotal (${extracted['subtotal'].value}) + tax (${extracted['tax'].value}) = "
                f"${cents_to_decimal_string(computed_total_cents)}, but the printed total is "
                f"${extracted['total'].value} — exceeds the ${tax_rounding_tolerance_usd:.2f} rounding tolerance."
            )

    controls.append(
        {
            "controlId": "arithmetic_tax_total",
            "label": "Tax and grand-total recalculation",
            "status": tax_status,
            "severity": "low" if tax_status == "passed" else "high",
            "reason": tax_reason,
            "evidenceReferences": ["tax", "total"],
            "blocking": True,
        }
    )

    return controls
